test/extensions/filters/http/custom_response/custom_response_integration_test.cc

```diff
@@ -0,0 +1,353 @@
+                                                       cer_position);
+              cer_position = hcm.mutable_http_filters()->size() - 2;
+            }
+            for (const auto& config : filters_after_cer_) {
+              auto* filter = hcm.mutable_http_filters()->Add();
+              TestUtility::loadFromYaml(config, *filter);
+              // swap with router filter
+              hcm.mutable_http_filters()->SwapElements(hcm.mutable_http_filters()->size() - 2,
+                                                       hcm.mutable_http_filters()->size() - 1);
+            }
+          }
+        });
+    HttpProtocolIntegrationTest::initialize();
+  }
+
+  CustomResponseIntegrationTest() : custom_response_filter_config_{kDefaultConfig} {}
+
+  void setPerRouteConfig(Route* route, const CustomResponse& cfg) {
+    Any cfg_any;
+    ASSERT_TRUE(cfg_any.PackFrom(cfg));
+    route->mutable_typed_per_filter_config()->insert(
+        MapPair<std::string, Any>("envoy.filters.http.custom_response", cfg_any));
+  }
+
+  void setPerHostConfig(VirtualHost& vh, const CustomResponse& cfg) {
+    Any cfg_any;
+    ASSERT_TRUE(cfg_any.PackFrom(cfg));
+    vh.mutable_typed_per_filter_config()->insert(
+        MapPair<std::string, Any>("envoy.filters.http.custom_response", cfg_any));
+  }
+
+protected:
+  Http::TestResponseHeaderMapImpl unauthorized_response_{{":status", "401"},
+                                                         {"content-length", "0"}};
+  Http::TestResponseHeaderMapImpl gateway_error_response_{{":status", "502"},
+                                                          {"content-length", "0"}};
+  Http::TestResponseHeaderMapImpl okay_response_{{":status", "201"}, {"content-length", "0"}};
+
+  Envoy::Http::LowerCaseString test_header_key_{kTestHeaderKey};
+  std::string custom_response_filter_config_;
+  std::vector<std::string> filters_before_cer_;
+  std::vector<std::string> filters_after_cer_;
+};
+
+// Verify that we get expected error response if the custom response is not configured.
+TEST_P(CustomResponseIntegrationTest, CustomResponseNotConfigured) {
+
+  // Use base class initialize.
+  HttpProtocolIntegrationTest::initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto response =
+      sendRequestAndWaitForResponse(default_request_headers_, 0, unauthorized_response_, 0);
+  EXPECT_TRUE(response->complete());
+  EXPECT_EQ("401", response->headers().getStatusValue());
+  EXPECT_EQ(1, test_server_->counter("http.config_test.downstream_rq_4xx")->value());
+}
+
+// Verify we get the correct local custom response.
+TEST_P(CustomResponseIntegrationTest, LocalReply) {
+  initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  default_request_headers_.setHost("some.route");
+  auto response =
+      sendRequestAndWaitForResponse(default_request_headers_, 0, unauthorized_response_, 0);
```

Comment:
The test name implies that Envoy's local reply is tested. But it looks like upstream fake is responding in this case.